File: packages/shared/src/agent/memory.js
'use strict';

/**
 * Typed, persistent workspace memory. Five scopes carry durable facts the
 * omnibox can save and recall:
 *   user      — per-user preferences, role, and stated facts
 *   business  — durable decisions about a business
 *   project   — decisions/notes scoped to a project
 *   task      — notes/outcomes scoped to a task or issue
 *   workspace — global notes alongside reviewed documentation
 *
 * This module is deterministic: validation, scope detection, and lexical recall
 * run with no model call (mirroring workspace-router.js / knowledge-search.js).
 * Persistence lives in store.js (addMemory/listMemories/...). Writes from free
 * text are surfaced as a *draft* and confirmed by the user before saving.
 */

const MAX_QUERY_CHARS = 8_000;
const MAX_TITLE_CHARS = 160;
const MAX_TEXT_CHARS = 2_000;
const MAX_TAGS = 8;
const MAX_TAG_CHARS = 40;
const MAX_RESULTS = 8;
const TITLE_FROM_TEXT_CHARS = 60;
const REFID_PATTERN = /^[A-Za-z0-9_-]{1,64}$/;

const MEMORY_SCOPES = Object.freeze(['user', 'business', 'project', 'task', 'workspace']);
const MEMORY_SOURCES = Object.freeze(['omnibox', 'business-pipeline', 'task', 'explicit']);

const STOP_WORDS = new Set([
  'about', 'after', 'also', 'and', 'are', 'can', 'check', 'find', 'for', 'from', 'have',
  'into', 'look', 'our', 'please', 'recall', 'remember', 'save', 'search', 'show', 'that',
  'the', 'their', 'this', 'what', 'when', 'where', 'which', 'with', 'workspace', 'would', 'your',
]);

class MemoryError extends Error {
  constructor(message, status = 400) {
    super(message);
    this.name = 'MemoryError';
    this.status = status;
  }
}

function cleanText(value, max) {
  return String(value == null ? '' : value).replace(/\s+/g, ' ').trim().slice(0, max);
}

/**
 * Boundary validation for a memory write. Returns ONLY the allowlisted fields —
 * an attacker-supplied `id`/`createdAt` cannot slip through (no mass assignment).
 */
function normalizeMemory(body) {
  if (!body || typeof body !== 'object' || Array.isArray(body)) {
    throw new MemoryError('A memory object is required.');
  }
  const scope = String(body.scope || '').trim().toLowerCase();
  if (!MEMORY_SCOPES.includes(scope)) {
    throw new MemoryError(`scope must be one of: ${MEMORY_SCOPES.join(', ')}.`);
  }
  const text = cleanText(body.text, MAX_TEXT_CHARS);
  if (!text) throw new MemoryError('text is required.');

  let refId = null;
  if (body.refId != null && body.refId !== '') {
    refId = String(body.refId);
    if (!REFID_PATTERN.test(refId)) {
      throw new MemoryError('refId may contain only letters, numbers, "_" and "-" (max 64).');
    }
  }

  const title = cleanText(body.title, MAX_TITLE_CHARS) || text.slice(0, TITLE_FROM_TEXT_CHARS);
  const tags = Array.isArray(body.tags)
    ? [...new Set(body.tags.map((tag) => cleanText(tag, MAX_TAG_CHARS)).filter(Boolean))].slice(0, MAX_TAGS)
    : [];
  const source = MEMORY_SOURCES.includes(body.source) ? body.source : 'omnibox';
  return { scope, refId, title, text, tags, source };
}

// Keyword fallbacks, checked in priority order after explicit "<scope> memory".
const SCOPE_KEYWORDS = Object.freeze([
  ['user', /\b(?:remember (?:that )?i\b|i (?:prefer|like|want|always|usually|am|work)\b|my (?:name|role|preference|email|timezone|style|goal)\b|about me|remind me|note to self|for me)\b/i],
  ['task', /\b(?:tasks?|issues?|tickets?|stor(?:y|ies)|bugs?|backlog|sprint)\b/i],
  ['project', /\bprojects?\b/i],
  ['business', /\b(?:business|revenue|pricing|prices?|customers?|markets?|moneti[sz]\w*|sales|profit|margins?|go-to-market|startup|company)\b/i],
  ['workspace', /\b(?:docs?|documentation|readme|guides?|how (?:do|to)|wiki)\b/i],
]);

/** Deterministic scope inference. Returns a scope or 'all' when ambiguous. */
function detectMemoryScope(query) {
  const input = cleanText(query, MAX_QUERY_CHARS);
  if (!input) return 'all';
  const explicit = input.match(/\b(user|business|project|task|workspace)\s+memor(?:y|ies)\b/i);
  if (explicit) return explicit[1].toLowerCase();
  for (const [scope, pattern] of SCOPE_KEYWORDS) {
    if (pattern.test(input)) return scope;
  }
  return 'all';
}

const WRITE_TRIGGER = /\b(?:remember(?:\s+that)?|note to self|keep in mind|make a note(?:\s+that)?)\b/i;
const SCOPED_WRITE = /\bsave (?:this )?(?:to|in) (?:my |our |the )?(user|business|project|task|workspace) memor(?:y|ies)\s*[:\-]?\s*(.*)$/i;

/**
 * Recognize a "remember this" style write request and return a DRAFT
 * ({ scope, title, text }) for the user to confirm — this never persists.
 * Returns null when the input is a plain read/query.
 */
function detectMemoryWrite(input) {
  const text = cleanText(input, MAX_QUERY_CHARS);
  if (!text) return null;

  const scoped = text.match(SCOPED_WRITE);
  if (scoped) {
    return draft(scoped[1].toLowerCase(), cleanText(scoped[2], MAX_TEXT_CHARS) || text);
  }
  if (!WRITE_TRIGGER.test(text)) return null;

  const body = cleanText(
    text.replace(/^.*?\b(?:remember(?:\s+that)?|note to self|keep in mind|make a note(?:\s+that)?)\s*[:\-]?\s*/i, ''),
    MAX_TEXT_CHARS,
  ) || text;
  return draft(detectMemoryScope(text), body);
}

function draft(scope, body) {
  const scopeOut = MEMORY_SCOPES.includes(scope) ? scope : 'workspace';
  const text = body || '';
  return { scope: scopeOut, title: text.slice(0, TITLE_FROM_TEXT_CHARS) || 'Saved note', text };
}

function queryTerms(query) {
  return [...new Set(cleanText(query, MAX_QUERY_CHARS).toLowerCase().match(/[a-z0-9][a-z0-9_-]{1,}/g) || [])]
    .filter((term) => term.length > 2 && !STOP_WORDS.has(term))
    .slice(0, 20);
}

function matchScore(text, terms) {
  const haystack = String(text || '').toLowerCase();
  return terms.reduce((score, term) => score + (haystack.includes(term) ? 1 : 0), 0);
}

/** Bounded lexical recall over stored memories, optionally filtered by scope. */
function searchMemories(query, memories, options = {}) {
  const scope = options.scope && options.scope !== 'all' ? options.scope : null;
  const terms = queryTerms(query);
  const pool = (Array.isArray(memories) ? memories : []).filter((memory) => !scope || memory.scope === scope);
  return pool
    .map((memory, index) => ({
      id: memory.id,
      scope: memory.scope,
      refId: memory.refId || null,
      type: `Memory · ${memory.scope}`,
      title: memory.title || 'Memory',
      summary: cleanText(memory.text, 320),
      status: memory.source || 'Saved',
      score: matchScore(`${memory.title} ${memory.text} ${(memory.tags || []).join(' ')}`, terms),
      index,
    }))
    .filter((record) => !terms.length || record.score > 0)
    .sort((a, b) => b.score - a.score || a.index - b.index)
    .slice(0, MAX_RESULTS)
    .map(({ index, ...record }) => record);
}

module.exports = {
  MAX_QUERY_CHARS,
  MAX_TEXT_CHARS,
  MAX_RESULTS,
  MEMORY_SCOPES,
  MEMORY_SOURCES,
  MemoryError,
  normalizeMemory,
  detectMemoryScope,
  detectMemoryWrite,
  searchMemories,
};

File: ai_fleet/agent/memory.py
"""Typed, persistent workspace memory (port of agent/memory.js).

Five scopes carry durable facts the omnibox can save and recall:
    user      — per-user preferences, role, and stated facts
    business  — durable decisions about a business
    project   — decisions/notes scoped to a project
    task      — notes/outcomes scoped to a task or issue
    workspace — global notes alongside reviewed documentation

This module is deterministic: validation, scope detection, and lexical recall
run with no model call. Persistence lives in the store; writes from free text
are surfaced as a *draft* and confirmed by the user before saving.

Port notes:
- Store/output dict keys stay camelCase (`refId`, `createdAt` cross the store /
  HTTP boundary) per the SPA contract.
- JS regexes use ASCII `\\w`/`\\b` semantics, so patterns are compiled with
  `re.ASCII` to mirror them exactly (inputs are pre-normalized to ASCII spaces
  by `_clean_text`, so `\\s` divergence is moot).
- `REFID_PATTERN` uses `\\Z` (not `$`) so a trailing newline cannot pass, matching
  JS `$` end-of-input semantics.
"""

from __future__ import annotations

import re

MAX_QUERY_CHARS = 8_000
MAX_TITLE_CHARS = 160
MAX_TEXT_CHARS = 2_000
MAX_TAGS = 8
MAX_TAG_CHARS = 40
MAX_RESULTS = 8
TITLE_FROM_TEXT_CHARS = 60
REFID_PATTERN = re.compile(r"^[A-Za-z0-9_-]{1,64}\Z")

MEMORY_SCOPES = ("user", "business", "project", "task", "workspace")
MEMORY_SOURCES = ("omnibox", "business-pipeline", "approval-gate", "task", "explicit")

STOP_WORDS = frozenset([
    "about", "after", "also", "and", "are", "can", "check", "find", "for", "from", "have",
    "into", "look", "our", "please", "recall", "remember", "save", "search", "show", "that",
    "the", "their", "this", "what", "when", "where", "which", "with", "workspace", "would", "your",
])


class MemoryError(Exception):
    """Boundary validation failure for a memory write/query (HTTP 400 by default).

    Shadows the Python builtin within this module intentionally; call sites use it
    module-qualified so there is no real collision.
    """

    def __init__(self, message, status=400):
        super().__init__(message)
        self.name = "MemoryError"
        self.message = message
        self.status = status


def _stringify(value):
    """Mirror JS `String(value)` for the scalar cases these boundaries see."""
    if value is True:
        return "true"
    if value is False:
        return "false"
    return str(value)


def _clean_text(value, max_):
    """`String(value ?? '').replace(/\\s+/g, ' ').trim().slice(0, max)`."""
    s = "" if value is None else _stringify(value)
    return re.sub(r"\s+", " ", s).strip()[:max_]


def normalize_memory(body):
    """Boundary validation for a memory write. Returns ONLY the allowlisted
    fields — an attacker-supplied `id`/`createdAt` cannot slip through (no mass
    assignment).
    """
    if not isinstance(body, dict):
        raise MemoryError("A memory object is required.")
    scope = str(body.get("scope") or "").strip().lower()
    if scope not in MEMORY_SCOPES:
        raise MemoryError(f"scope must be one of: {', '.join(MEMORY_SCOPES)}.")
    text = _clean_text(body.get("text"), MAX_TEXT_CHARS)
    if not text:
        raise MemoryError("text is required.")

    ref_id = None
    if body.get("refId") is not None and body.get("refId") != "":
        ref_id = _stringify(body.get("refId"))
        if not REFID_PATTERN.match(ref_id):
            raise MemoryError('refId may contain only letters, numbers, "_" and "-" (max 64).')

    title = _clean_text(body.get("title"), MAX_TITLE_CHARS) or text[:TITLE_FROM_TEXT_CHARS]

    tags = []
    if isinstance(body.get("tags"), list):
        seen = []
        for tag in body["tags"]:
            cleaned = _clean_text(tag, MAX_TAG_CHARS)
            if cleaned and cleaned not in seen:
                seen.append(cleaned)
        tags = seen[:MAX_TAGS]

    source = body.get("source") if body.get("source") in MEMORY_SOURCES else "omnibox"
    return {"scope": scope, "refId": ref_id, "title": title, "text": text, "tags": tags, "source": source}


# Keyword fallbacks, checked in priority order after explicit "<scope> memory".
SCOPE_KEYWORDS = (
    ("user", re.compile(
        r"\b(?:remember (?:that )?i\b|i (?:prefer|like|want|always|usually|am|work)\b|"
        r"my (?:name|role|preference|email|timezone|style|goal)\b|about me|remind me|"
        r"note to self|for me)\b",
        re.IGNORECASE | re.ASCII)),
    ("task", re.compile(
        r"\b(?:tasks?|issues?|tickets?|stor(?:y|ies)|bugs?|backlog|sprint)\b",
        re.IGNORECASE | re.ASCII)),
    ("project", re.compile(r"\bprojects?\b", re.IGNORECASE | re.ASCII)),
    ("business", re.compile(
        r"\b(?:business|revenue|pricing|prices?|customers?|markets?|moneti[sz]\w*|sales|"
        r"profit|margins?|go-to-market|startup|company)\b",
        re.IGNORECASE | re.ASCII)),
    ("workspace", re.compile(
        r"\b(?:docs?|documentation|readme|guides?|how (?:do|to)|wiki)\b",
        re.IGNORECASE | re.ASCII)),
)

_EXPLICIT_SCOPE = re.compile(
    r"\b(user|business|project|task|workspace)\s+memor(?:y|ies)\b",
    re.IGNORECASE | re.ASCII)


def detect_memory_scope(query):
    """Deterministic scope inference. Returns a scope or 'all' when ambiguous."""
    input_ = _clean_text(query, MAX_QUERY_CHARS)
    if not input_:
        return "all"
    explicit = _EXPLICIT_SCOPE.search(input_)
    if explicit:
        return explicit.group(1).lower()
    for scope, pattern in SCOPE_KEYWORDS:
        if pattern.search(input_):
            return scope
    return "all"


WRITE_TRIGGER = re.compile(
    r"\b(?:remember(?:\s+that)?|note to self|keep in mind|make a note(?:\s+that)?)\b",
    re.IGNORECASE | re.ASCII)
SCOPED_WRITE = re.compile(
    r"\bsave (?:this )?(?:to|in) (?:my |our |the )?"
    r"(user|business|project|task|workspace) memor(?:y|ies)\s*[:\-]?\s*(.*)$",
    re.IGNORECASE | re.ASCII)
_WRITE_STRIP = re.compile(
    r"^.*?\b(?:remember(?:\s+that)?|note to self|keep in mind|make a note(?:\s+that)?)\s*[:\-]?\s*",
    re.IGNORECASE | re.ASCII)


def detect_memory_write(input_):
    """Recognize a "remember this" style write request and return a DRAFT
    ({scope, title, text}) for the user to confirm — this never persists.
    Returns None when the input is a plain read/query.
    """
    text = _clean_text(input_, MAX_QUERY_CHARS)
    if not text:
        return None

    scoped = SCOPED_WRITE.search(text)
    if scoped:
        return _draft(scoped.group(1).lower(), _clean_text(scoped.group(2), MAX_TEXT_CHARS) or text)
    if not WRITE_TRIGGER.search(text):
        return None

    body = _clean_text(_WRITE_STRIP.sub("", text, count=1), MAX_TEXT_CHARS) or text
    return _draft(detect_memory_scope(text), body)


def _draft(scope, body):
    scope_out = scope if scope in MEMORY_SCOPES else "workspace"
    text = body or ""
    return {"scope": scope_out, "title": text[:TITLE_FROM_TEXT_CHARS] or "Saved note", "text": text}


def _query_terms(query):
    lowered = _clean_text(query, MAX_QUERY_CHARS).lower()
    matches = re.findall(r"[a-z0-9][a-z0-9_-]{1,}", lowered)
    seen = []
    for term in matches:
        if term not in seen:
            seen.append(term)
    return [term for term in seen if len(term) > 2 and term not in STOP_WORDS][:20]


def _match_score(text, terms):
    haystack = _stringify(text or "").lower()
    return sum(1 for term in terms if term in haystack)


def _template(value):
    """Mirror a JS template literal `${value}`: undefined/None -> 'undefined'."""
    if value is None:
        return "undefined"
    return _stringify(value)


def _join(arr):
    """Mirror `Array.prototype.join(' ')`: null/undefined elements -> ''."""
    return " ".join("" if x is None else _stringify(x) for x in arr)


def search_memories(query, memories, options=None):
    """Bounded lexical recall over stored memories, optionally filtered by scope."""
    options = options or {}
    opt_scope = options.get("scope")
    scope = opt_scope if (opt_scope and opt_scope != "all") else None
    terms = _query_terms(query)
    pool = memories if isinstance(memories, list) else []
    pool = [m for m in pool if (not scope or m.get("scope") == scope)]

    scored = []
    for index, memory in enumerate(pool):
        title_raw = memory.get("title")
        text_raw = memory.get("text")
        scope_raw = memory.get("scope")
        tags_raw = memory.get("tags") or []
        score = _match_score(
            f"{_template(title_raw)} {_template(text_raw)} {_join(tags_raw)}", terms)
        record = {
            "id": memory.get("id"),
            "scope": scope_raw,
            "refId": memory.get("refId") or None,
            "type": f"Memory · {_template(scope_raw)}",
            "title": title_raw or "Memory",
            "summary": _clean_text(text_raw, 320),
            "status": memory.get("source") or "Saved",
            "score": score,
        }
        scored.append((score, index, record))

    scored = [entry for entry in scored if (not terms or entry[0] > 0)]
    scored.sort(key=lambda entry: (-entry[0], entry[1]))
    return [record for (_score, _index, record) in scored[:MAX_RESULTS]]
